### ISSUE TITLE
Flutter Web truncates logs written with dart:developer

### ISSUE THREAD

**Issue by tomasweigenast** [flutter/devtools#3222]
Its simple, Flutter Web truncates log messages written to the console using `dart:developer`. Something that should not happen.

This is what gets printed:
```[LOG] [26/03/2021 19:35:19] [DEBUG] Sending POST request…ccount/login with parameters null. Body: true```

Can you notice the `...` in the middle of the text?

If I log the same line in a mobile app, that does not happen

Flutter doctor:
```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel beta, 2.1.0-12.2.pre, on Microsoft Windows [VersiÃ³n 10.0.18363.1440], locale es-AR)
[!] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    X Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.
[√] Chrome - develop for the web
[√] Android Studio (version 4.0)
[√] VS Code (version 1.54.3)
[√] Connected device (3 available)

! Doctor found issues in 1 category.
```

**Comment by nehal076** [comment]
Any update on this? I'm facing the same issue. Logs are clipped. I don't know why.

![image](https://user-images.githubusercontent.com/32421196/152147231-5c156471-e169-40f2-9210-958c44ad4d87.png)

**Comment by jacob314** [comment]
@annagrin you were looking at batching events for dart:developer dispatchEvent. Can we use the same batching to resolve this truncation case?

**Comment by nshahan** [comment]
@nehal076 What application is being shown in your screenshot? From the perspective of the dart:developer library we are not truncating any text being logged. From what I have seen the truncation is performed by the UI of some applications and is only changing what is being displayed, not the actual contents of the log message. I have observed that I can see the entire message by hovering or clicking and expanding.

**Comment by nshahan** [comment]
@jacob314 @annagrin If we are copying the output from the chrome console and showing it in devtools it is possible we need to find a way to get the full log message, not just the text the UI of the console is printing. The text displayed in the chrome console can be truncated.

**Comment by nehal076** [comment]
> @nehal076 What application is being shown in your screenshot? From the perspective of the dart:developer library we are not truncating any text being logged. From what I have seen the truncation is performed by the UI of some applications and is only changing what is being displayed, not the actual contents of the log message. I have observed that I can see the entire message by hovering or clicking and expanding. 

Hello.. 
This is VSCode, debug console. It works fine when I run the app for mobile. Only clips the logs for web.

**Comment by jacob314** [comment]
Fyi @DanTup. Perhaps the issue is that on mobile we automatically wrap lines in Flutter tools but on the web we are not resulting in truncated VSCode output. If it is a wrapping issue, we might want to simulate wrapping in DWDS so behavior is consistent between the VM and Web.

**Comment by DanTup** [comment]
I don't know if it helps, but here's how the events that VS Code gets look. For iOS simulator, the event's `logRecord.message` is truncated at 128 characters and `valueAsStringIsTruncated` is set to `true`, which causes VS Code to call `getObject` on it and get the full string:

```js
{
	"jsonrpc": "2.0",
	"method": "streamNotify",
	"params": {
		"streamId": "Logging",
		"event": {
			"type": "Event",
			"kind": "Logging",
			"isolate": {
				"type": "@Isolate",
				"id": "isolates/1656465624059587",
				"name": "main",
				"number": "1656465624059587",
				"isSystemIsolate": false
			},
			"timestamp": 1647277968461,
			"logRecord": {
				// ...
				"message": {
					// ...
					"identityHashCode": 0,
					"kind": "String",
					"id": "objects/1",
					"length": 1000,
					"valueAsString": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001111111111111111111111111111",
					"count": 128,
					"valueAsStringIsTruncated": true
				},
```

When running on Chrome, the event is truncated to 100 characters (though in the middle) and does not have `valueAsStringIsTruncated` so we just show the `valueAsString` value:

```js
{
	"jsonrpc": "2.0",
	"method": "streamNotify",
	"params": {
		"streamId": "Logging",
		"event": {
			"type": "Event",
			"kind": "Logging",
			"timestamp": 1647278009535,
			"isolate": {
				"type": "@Isolate",
				"id": "1",
				"number": "1",
				"name": "main()",
				"isSystemIsolate": false
			},
			"logRecord": {
				"type": "LogRecord",
				"message": {
					"type": "@Instance",
					"id": "#StringInstanceRef#00000000000000000000000000000000000000000000000000…9999999999999999999999999999999999999999999999999",
					"kind": "String",
					"identityHashCode": 669583697,
					// ...
					"valueAsString": "00000000000000000000000000000000000000000000000000…9999999999999999999999999999999999999999999999999",
					"length": 100
				},
```

**Comment by annagrin** [comment]
The truncated log is communicated by chrome to dwds. 

> We changed two top level methods from the dart:developer (postEvent() and registerExtension()) to use a different messaging channel so those messages are no longer being printed to the console using console.debug(). This means nothing will appear in the browser console for these APIs. Instead they appear in Flutter/Dart the DevTools console.

@nshahan I think it should be possible to use the same mechanism for `log()` as well.

**Comment by motucraft** [comment]
Any updates?

**Comment by lucasdidur** [comment]
Up

**Comment by icecandy** [comment]
I'm using Android Studio and the logging is still truncated using DartDeveloper.log()

**Comment by jacobsimionato** [comment]
This is affecting at least one Google app who would like to access well-organised and non-truncated log messages in the DevTools console. I'm curious what this is blocked on @annagrin - is it SDK work to change the channel used for log messages as @nshahan suggested? Would it be helpful to file a separate issue for this?

**Comment by elliette** [comment]
Just filed a separate issue so that we can track this work on the DWDS (and potentially Dart SDK) side.

**Comment by nshahan** [comment]
@annagrin and I discussed this today and have worked out a path forward (please feel free to comment on or update these steps if I didn't get it quite right).

1) Update the Dart SDK to test for the existence of a callback. If it exists it will be called with the log message. If not the message will be printed with `console.debug()` as it is today. Include a TODO to eventually remove this fallback and replace with a warning message listing the required version of dwds (See step 5).
2) Wait for a Dart SDK dev release to be published with the change.
3) Update the appropriate packages used by dwds (might be itself or separate packages for external vs google3 usage) to register the callback when the app starts up. 
4) Publish dwds with a lower SDK bound that includes the change.
5) Update Dart SDK to remove the fallback and instead print a warning message with the required version of dwds.
6) Update dwds dep in Flutter and roll into google3.

@jacobsimionato @annagrin 
Since we are going through this process, there is an open question: Are there any other dart:developer APIs we should update in the same way? The one we are unsure about is `inspect()` https://api.flutter.dev/flutter/dart-developer/inspect.html

**Comment by jacobsimionato** [comment]
Fantastic - this is great! Thanks so much, Dart team!

Re the `inspect()` API: I'm not sure if this helps, but I put some internal Google context about usage in b/319337383. Basically, it seems like inspect() isn't used in Google internal code. Though if it's broken in some way and there's an opportunity to fix it now, maybe that makes sense?

**Comment by elliette** [comment]
The `dart:developer` log changes have been rolled into google3, therefore marking this as fixed.

I did some investigation, and I don't think we need to do the same thing for the `inspect` API: https://github.com/dart-lang/webdev/issues/2335
